classes/webservice/WebserviceSpecificManagementAttachments.php

```diff
@@ -0,0 +1,316 @@
+        if (!$attachment) {
+            throw new WebserviceException(
+                sprintf(
+                    'Attachment %d not found',
+                    $this->wsObject->urlSegment[2]
+                ),
+                [
+                    1,
+                    Response::HTTP_INTERNAL_SERVER_ERROR,
+                ]
+            );
+        }
+
+        // Physical file location
+        $file = _PS_DOWNLOAD_DIR_ . $attachment->file;
+        // Check if file exists
+        if (!file_exists($file)) {
+            throw new WebserviceException(
+                sprintf(
+                    'Unable to load the attachment file for attachment %d',
+                    $this->wsObject->urlSegment[2]
+                ),
+                [
+                    1,
+                    Response::HTTP_INTERNAL_SERVER_ERROR,
+                ]
+            );
+        }
+
+        // Return file details
+        return array(
+            'file' => $file,
+            'mime' => $attachment->mime,
+            'file_name' => $attachment->file_name,
+            'file_size' => $attachment->file_size,
+        );
+    }
+
+    /**
+     * Handles file upload
+     *
+     * Creates new attachment or replaces existing with a new file.
+     * [PUT] update existing attachment file
+     * [POST] create new attachment
+     */
+    public function executeFileAddAndEdit()
+    {
+        // Load attachment with or without id depending on method
+        $attachment = new Attachment($this->wsObject->method == 'PUT' ? (int) $this->wsObject->urlSegment[1] : null);
+
+        // Check form data
+        if (isset($_FILES['file']) && is_uploaded_file($_FILES['file']['tmp_name'])) {
+            // Ensure file is within allowed size limit
+            if ($_FILES['file']['size'] > (Configuration::get('PS_ATTACHMENT_MAXIMUM_SIZE') * 1024 * 1024)) {
+                $this->wsObject->errors[] = sprintf(
+                    $this->l('The file is too large. Maximum size allowed is: %1$d kB. The file you are trying to upload is %2$d kB.'),
+                    Configuration::get('PS_ATTACHMENT_MAXIMUM_SIZE') * 1024,
+                    number_format(($_FILES['file']['size'] / 1024), 2, '.', '')
+                );
+            } else {
+                // Assign unique id
+                $uniqid = sha1(uniqid()); // must be a sha1
+
+                $attachment->file_name = $_FILES['file']['name'];
+                $attachment->file = $uniqid;
+                $attachment->mime = $_FILES['file']['type'];
+                $attachment->name[Configuration::get('PS_LANG_DEFAULT')] = $_POST['name'];
+
+                // Move file to download dir
+                if (!move_uploaded_file($_FILES['file']['tmp_name'], _PS_DOWNLOAD_DIR_ . $uniqid)) {
+                    $this->wsObject->errors[] = $this->l('Failed to copy the file.');
+                    unlink(_PS_DOWNLOAD_DIR_ . $attachment->file);
+                    if ($attachment->id) {
```

Comment:
Ok, on second thought, only the error assign line should be here.
If a fail occurs for a new attachment, no attachment will be persisted to database.
If a fail occurs for an existing attachment, nothing should be done.

So remove all of this (the file should not be there, and previous attachment info should be kept)
                    **unlink(_PS_DOWNLOAD_DIR_ . $attachment->file);
                    if ($attachment->id) {
                        $attachment->delete();
                    }**

Reply:
Thanks for explanation!